File: app/src/main/java/top/linl/qstorycloud/hook/update/UpdateChecker.java
package top.linl.qstorycloud.hook.update;

import android.content.ContentResolver;
import android.database.Cursor;
import android.net.Uri;

import com.alibaba.fastjson2.JSON;

import java.util.Timer;
import java.util.TimerTask;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.ObservableOnSubscribe;
import io.reactivex.rxjava3.disposables.Disposable;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import top.linl.qstorycloud.db.LocalModuleInfoDAO;
import top.linl.qstorycloud.db.helper.UpdateDBHelper;
import top.linl.qstorycloud.hook.HookEnv;
import top.linl.qstorycloud.hook.moduleloader.model.LocalModuleInfo;
import top.linl.qstorycloud.hook.update.model.UpdateInfo;
import top.linl.qstorycloud.hook.util.ToastTool;
import top.linl.qstorycloud.log.QSLog;

public class UpdateChecker {
    private final String detectUpdatesUrl = "https://qstory.linl.top/update/detectUpdates";

    private UpdateInfo lastUpdateInfo;

    private void run() {
        //这里用到了跨进程通讯
        //获取内容提供者
        ContentResolver contentResolver = HookEnv.getHostAppContext().getContentResolver();
        //查询是否开启了安全模式
        Uri uri = Uri.parse("content://qstorycloud.linl.top/cleanData");
        Cursor cursor = contentResolver.query(uri, new String[]{"value"}, "name=?", new String[]{"clean_data"}, null);
        String isCleanData = null;
        if (cursor != null && cursor.moveToNext()) {
            isCleanData = cursor.getString(0);
            cursor.close();
        }
        if (isCleanData != null) {
            UpdateDBHelper dbHelper = UpdateDBHelper.getInstance();
            dbHelper.onUpgrade(dbHelper.getWritableDatabase(), 1, 1);
            ToastTool.show("收到重置指令");
        }
    }
    public void startObservingUpdates() {
        if (!HookEnv.isMainProcess()) {
            return;
        }
        run();
        //观察者
        Disposable disposable = Observable.create((ObservableOnSubscribe<UpdateInfo>) emitter -> {
            //先去查询一次之前的版本
            //循环检测更新
            Timer timer = new Timer();
            timer.schedule(new TimerTask() {
                @Override
                public void run() {
                    UpdateInfo updateInfo = getUpdateInfo();
                    if (updateInfo == null) return;
                    if (updateInfo.equals(lastUpdateInfo)) return;
                    lastUpdateInfo = updateInfo;
                    emitter.onNext(updateInfo);
                }
            }, 0, 10 * 1000);//每十分钟轮询一次
        }).subscribe(new UpdateObserver());
    }

    /**
     * 请求最新版本
     */
    private UpdateInfo getUpdateInfo() {
        LocalModuleInfo localModuleInfo = LocalModuleInfoDAO.getLastModuleInfo();
        int moduleVersionCode = 0;
        if (localModuleInfo != null) {
            moduleVersionCode = localModuleInfo.getModuleVersionCode();
        }
        OkHttpClient client = new OkHttpClient.Builder().build();
        FormBody formBody = new FormBody.Builder()
                .add("versionCode", String.valueOf(moduleVersionCode))
                .build();
        Request request = new Request
                .Builder()
                .url(detectUpdatesUrl)
                .post(formBody)
                .addHeader("User-Agent", "Android")
                .addHeader("Content-Type", "text/plain")
                .addHeader("Accept", "*/*")
                .addHeader("Connection", "keep-alive")
                .build();
        try (Response response = client.newCall(request).execute()) {
            String data = response.body().string();
            return JSON.parseObject(data, UpdateInfo.class);
        } catch (Exception e) {
            QSLog.e("DetectUpdates", e);
            return null;
        }

    }
}

File: app/src/main/java/top/linl/qstorycloud/hook/update/UpdateChecker.kt
package top.linl.qstorycloud.hook.update

import kotlinx.coroutines.flow.FlowCollector
import top.linl.qstorycloud.config.LocalModuleData
import top.linl.qstorycloud.config.UpdateInfoData
import top.linl.qstorycloud.hook.HookEnv
import top.linl.qstorycloud.hook.PathTool
import top.linl.qstorycloud.hook.util.ToastTool
import top.linl.qstorycloud.log.QSLog
import top.linl.qstorycloud.model.LocalModuleInfo
import top.linl.qstorycloud.model.UpdateInfo
import top.linl.qstorycloud.util.SpHelper
import java.io.File

class UpdateChecker : FlowCollector<UpdateInfo> {

    val spHelper = SpHelper("UpdateChecker")

    override suspend fun emit(value: UpdateInfo) {
        QSLog.d(this, "UpdateInfo: $value")

        val lastModuleInfo = LocalModuleData.getLastModuleInfo()
        val lastUpdateInfo = UpdateInfoData.getLastUpdateInfo()
        //如果从未初始化过
        if (lastModuleInfo == null || lastUpdateInfo == null || !File(lastModuleInfo.moduleApkPath).exists()) {
            QSLog.d(this, "未初始化过last module:$lastModuleInfo, update info: $lastUpdateInfo")
            download(value)
            return
        }

        //如果云端版本大于本地更新版本
        if (value.latestVersionCode > lastUpdateInfo.latestVersionCode) {
            QSLog.d(this, "云端版本大于本地版本")
            download(value)
            return
        }

        //如果云端版本大于 本地版本
        if (lastUpdateInfo.latestVersionCode > lastModuleInfo.moduleVersionCode) {
            download(value)
            return
        }
    }

    private fun download(updateInfo: UpdateInfo) {
        UpdateInfoData.addUpdateInfo(updateInfo)
        QSLog.d(this, "开始下载")
        ToastTool.show("QStory开始更新")
        //下载
        val downloadTask = DownloadTask(
            HookEnv.getHostAppContext(),
            updateInfo
        )
        val downloadPath = PathTool.getDataPath() + "/" + updateInfo.latestVersionName
        downloadTask.download(updateInfo.updateUrl, downloadPath)
        //下载完成 初始化bean
        val localModuleInfo = LocalModuleInfo().apply {
            moduleApkPath = downloadPath
            moduleName = updateInfo.latestVersionName
            moduleVersionCode = updateInfo.latestVersionCode
            moduleVersionName = updateInfo.latestVersionName
        }
        LocalModuleData.addModuleInfo(localModuleInfo)
        ToastTool.show("QStory已更新完成，请重启QQ")
        QSLog.d(this, "下载完成")
    }
}